core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PushFilterIntoTableScanWithPredicatePruning.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.optimizations;
+
+import io.trino.Session;
+import io.trino.cost.CachingStatsProvider;
+import io.trino.cost.StatsCalculator;
+import io.trino.cost.StatsProvider;
+import io.trino.cost.TableStatsProvider;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.planner.DomainTranslator;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.SymbolAllocator;
+import io.trino.sql.planner.TypeAnalyzer;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.iterative.rule.PushPredicateIntoTableScan;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.SimplePlanRewriter;
+import io.trino.sql.planner.plan.TableScanNode;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * The difference between this optimizer and {@link PushPredicateIntoTableScan} rule is that we
+ * invoke {@link PushPredicateIntoTableScan#pushFilterIntoTableScan} here with pruneWithPredicateExpression=true.
+ */
+public class PushFilterIntoTableScanWithPredicatePruning
```

Comment:
We shouldn't have to have a separate optimizer for this. It can be done by parameterizing the PushPredicateIntoTableScan to optionally prune with predicate expression.

Also, it should be implemented as a Rule, not as a PlanOptimizer -- we're trying to deprecate visitor based optimizers in favor of rules wherever possible.

Reply:
@martint added another invocation of PushPredicateIntoTableScan iterative optimizer